lib/core/src/Cardano/Wallet/Primitive/Types/TokenMap.hs

```diff
@@ -0,0 +1,529 @@
+import GHC.Generics
+    ( Generic )
+import GHC.TypeLits
+    ( ErrorMessage (..), TypeError )
+import Quiet
+    ( Quiet (..) )
+
+import qualified Cardano.Wallet.Primitive.Types.TokenQuantity as TQ
+import qualified Data.Aeson as Aeson
+import qualified Data.Foldable as F
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.NonEmpty.Strict as NEMap
+import qualified Data.Map.Strict as Map
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+-- | A map of named token quantities, grouped by token policy.
+--
+-- The token map data structure has an important invariant: all token
+-- quantities held within a map are non-zero.
+--
+-- This means that:
+--
+--   - using the 'setQuantity' function to add a zero-valued quantity to a
+--     map is equivalent to applying the identity operation to that map.
+--
+--   - using the 'setQuantity' function to change an existing quantity to zero
+--     is equivalent to removing that quantity from the map.
+--
+-- As a consequence of this invariant, the token map data structure is
+-- always in its canonical form: we can perform an equality check without
+-- needing any extra canonicalization steps.
+--
+newtype TokenMap = TokenMap
+    { unTokenMap
+        :: Map TokenPolicyId (NonEmptyMap TokenName TokenQuantity)
+    }
+    deriving stock (Eq, Generic)
+    deriving Show via (Quiet TokenMap)
+
+-- | Token maps can be partially ordered, but there is no total ordering of
+--   token maps that's consistent with their arithmetic properties.
+--
+-- In the event that someone attempts to define an 'Ord' instance for the
+-- 'TokenMap' type, we generate a type error.
+--
+-- If some arbitrary ordering is needed (for example, so that token maps can
+-- be included in an ordered set), the recommended course of action is to
+-- define a newtype with its own dedicated 'Ord' instance.
+--
+instance TypeError ('Text "Ord not supported for token maps")
+        => Ord TokenMap where
+    compare = error "Ord not supported for token maps"
+
+instance NFData TokenMap
+
+instance Semigroup TokenMap where
+    (<>) = add
+
+instance Monoid TokenMap where
+    mempty = empty
+
+-- | A combination of a token policy identifier and a token name that can be
+--   used as a compound identifier.
+--
+data AssetId = AssetId
+    { tokenPolicyId
+        :: !TokenPolicyId
+    , tokenName
+        :: !TokenName
+    }
+    deriving stock (Eq, Generic, Ord, Show)
+
+--------------------------------------------------------------------------------
+-- Serialization
+--------------------------------------------------------------------------------
+
+-- | When used with the 'Buildable' or 'ToJSON' instances, provides a flat
+-- serialization style, where token quantities are paired with their asset
+-- identifiers.
+--
+newtype Flat a = Flat { getFlat :: a }
```

Comment:
Are both of Flat and Nested _really_ needed?
I would expect the JSON serialisation to be defined only in the Api.Types module, and according to the swagger spec.

Reply:
> Are both of Flat and Nested really needed?

Good point.

We could easily get rid of one if we can show that only one is useful.

`Nested` is useful for debugging, as it groups tokens together by policy, saving space. It might also be useful for clients such as Daedalus, to provide a view that groups tokens by policy (since tokens can share the same name). But we should check this before deciding what to eliminate.

> I would expect the JSON serialisation to be defined only in the Api.Types module

I had anticipated that these types (and their JSON instances) would be re-used by several types within the `Api.Types` module.

Though for the `TokenMap` `FromJSON` instance, we perform validation to guard against zero values, which is something that `TokenMap` itself is responsible for.

We could move these to the `Api.Types` module, but then the responsibility for guarding against zero values would be located in two modules, rather than one. It seems better to put all the validation for a type together with the type, as this way they can also be tested together.

It's actually not the first time we've done this. The following types all define JSON instances in their own modules, in some cases for similar reasons:

- `Quantity`
- `Percentage`
- `StakePoolTicker`
- `PoolOwner`
- `PoolMetadataSource`
- `SMASHPoolId`

Reply:
Can we try to keep the code limited to what is essential for purpose, rather than adding functions which "might be needed"?
For a start, that would make less code to review, and therefore less possible objections from me.
But also it helps reduce confusion when it comes time to refactor.

Reply:
> Can we try to keep the code limited to what is essential for purpose, rather than adding functions which "might be needed"?

As mentioned above, I'd be more than happy to remove one of these formats.

My original motivation for the flat/nested distinction was testing: visual feedback of grouping was useful while implementing these types.

I've asked for feedback from the Daedalus team about which JSON format they'd prefer for parsing (nested vs flat).